statistics/table.go

```diff
@@ -466,13 +450,70 @@ func getOrdinalOfRangeCond(sc *stmtctx.StatementContext, ran *ranger.Range) int
 	return len(ran.LowVal)
 }
 
-func (coll *HistColl) getIndexRowCount(sc *stmtctx.StatementContext, idx *Index, indexRanges []*ranger.Range, modifyCount int64) (float64, error) {
+// GenerateHistCollFromColumnInfo generates a new HistColl whose ColID2IdxID and IdxID2ColIDs is built from the given parameter.
+func (coll *HistColl) GenerateHistCollFromColumnInfo(infos []*model.ColumnInfo, columns []*expression.Column) HistColl {
+
+	newColHistMap := make(map[int64]*Column)
+	colInfoID2UniqueID := make(map[int64]int64)
+	colNames2UniqueID := make(map[string]int64)
+	for _, col := range columns {
+		colInfoID2UniqueID[col.ID] = col.UniqueID
+	}
+	for _, colInfo := range infos {
+		uniqueID, ok := colInfoID2UniqueID[colInfo.ID]
+		if ok {
+			colNames2UniqueID[colInfo.Name.L] = uniqueID
+		}
+	}
+	for id, colHist := range coll.Columns {
+		uniqueID, ok := colInfoID2UniqueID[id]
+		// If this column is not in datasource's schema, it won't be used in this query.
+		if ok {
+			newColHistMap[uniqueID] = colHist
```

Comment:
Do we need to adjust [here](https://github.com/pingcap/tidb/blob/master/statistics/table.go#L355), because the `colID` is unique id, not column id?

Reply:
emm... Let me think a while